app-ng/src/server.lisp

```diff
@@ -0,0 +1,135 @@
+;;;;
+;;;; Author: Robert Smith
+
+(in-package #:qvm-app-ng)
+
+(alexandria:define-constant +default-server-address+ "0.0.0.0"
+  :test #'string=
+  :documentation "The default host address on which the HTTP server will listen.")
+
+(alexandria:define-constant +default-server-port+ 5000
+  :documentation "The default port on which the HTTP server will listen.")
+
+(defclass rpc-acceptor (tbnl:easy-acceptor)
+  ()
+  (:default-initargs
+   :address (error "Host address must be specified.")
+   :document-root nil
+   :error-template-directory nil
+   :persistent-connections-p t))
+
+(defvar *rpc-acceptor* nil
+  "*RPC-ACCEPTOR* holds a reference to the RPC-ACCEPTOR instance created by last invocation of START-SERVER.")
+
+(defun start-server-mode (&key host port)
+  "Start the HTTP server on the indicated HOST and PORT. Does not return."
+  (check-type host string)
+  ;; A PORT of 0 tells hunchentoot to pick a random port.
+  (check-type port (or null (integer 0 65535)) "The port must be between 0 and 65535.")
+  (format-log "Starting server on port ~D." port)
+  (start-server host port)
+  (loop (sleep 60)))
+
+(defun start-server (host port &optional debug)
+  "Start the HTTP server on the indicated HOST and PORT.
+
+If the optional DEBUG is non-NIL, show additional debug info and enter the debugger on error.
+
+Returns the newly created and running RPC-ACCEPTOR, which is also saved in *RPC-ACCEPTOR* for REPL-debugging convenience."
+  (setf tbnl:*log-lisp-backtraces-p* debug
+        tbnl:*log-lisp-errors-p* debug
+        tbnl:*show-lisp-errors-p* debug
+        tbnl:*show-lisp-backtraces-p* debug
+        tbnl:*catch-errors-p* (not debug))
+  (setf *rpc-acceptor* (make-instance
+                        'rpc-acceptor
+                        :address host
+                        :port port
+                        :taskmaster (make-instance 'tbnl:one-thread-per-connection-taskmaster)))
+  (pushnew #'dispatch-rpc-handlers tbnl:*dispatch-table*)
+  (tbnl:reset-session-secret)
+  (tbnl:start *rpc-acceptor*))
+
+(defun stop-server (&optional (acceptor *rpc-acceptor*))
+  (tbnl:stop acceptor))
+
+(defun session-info ()
+  "Return a string describing the server session info."
+  (if (or (not (boundp 'tbnl:*session*))
+          (null tbnl:*session*))
+      ""
+      (format nil
+              "[~A Session:~D] "
+              (tbnl:session-remote-addr tbnl:*session*)
+              (tbnl:session-id tbnl:*session*))))
+
+(defun encode-json (object)
+  "Encode OBJECT to a JSON string."
+  (when (boundp 'tbnl:*reply*)
+    (setf (tbnl:content-type*) "application/json; charset=utf-8"))
+  (with-output-to-string (s)
+    (yason:encode object s)))
+
+(defun error-response (message)
+  "Return a JSON string of structured error MESSAGE of type qvm_error."
+  (encode-json
+   (alexandria:plist-hash-table
+    (list "error_type" "qvm_error"
+          "status" message))))
+
+(defvar *request-json*)
+(setf (documentation '*request-json* 'variable)
```

Comment:
I'm curious: why not set the documentation directly within the `defvar` form in this case? I've seen this done in SICL (adding all the docstrings in a separate lisp file) for clarity and internationalization purposes.

Reply:
Because then I'd have to provide a value, and I want it to unbound in the global environment.

Unless you mean something like

```
(setf (documentation (defvar *request-json*) 'variable) "...")
```

which I suppose is possible, but I think I prefer separate toplevel forms.

Reply:
Ah! I somehow interpolated a `nil` where there was none (no pun intended).

Reply:
In previous iteration of this PR, I did provide an initial-value of NIL (and documentation in the same `defvar` form). But I changed it to unbound after a good call-out from @notmgsk.